File: go/spore-core/tools/memory_test.go
package tools

import (
	"context"
	"encoding/json"
	"errors"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"testing"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
	"github.com/squirrelsoft-dev/spore-core/go/spore-core/storage"
)

// ── Test helpers ─────────────────────────────────────────────────────────────

func memCall(input map[string]any) sporecore.ToolCall {
	b, _ := json.Marshal(input)
	return sporecore.ToolCall{ID: "c1", Name: MemoryToolName, Input: b}
}

// ctxWith builds a ToolContext over a memory store keyed by session.
func ctxWith(store storage.MemoryStore, session string) *sporecore.ToolContext {
	return sporecore.NewToolContext(sporecore.SessionID(session), nil, store)
}

// inMemoryCtx is a fresh single-backend ToolContext. A single InMemory backend
// is scope-aware, so user and project entries stay isolated.
func inMemoryCtx() *sporecore.ToolContext {
	return ctxWith(storage.NewInMemoryStorageProvider(), "test-session")
}

// scopedCtx builds a ToolContext over a CompositeStorageProvider routing User /
// Project (and optionally Local) to their own InMemory backends — the canonical
// multi-scope wiring whose ScopedMemoryRouter owns the merge.
func scopedCtx(session string, withLocal bool) *sporecore.ToolContext {
	b := storage.NewCompositeStorageProvider().
		Memory(storage.StorageScopeUser, storage.NewInMemoryStorageProvider()).
		Memory(storage.StorageScopeProject, storage.NewInMemoryStorageProvider())
	if withLocal {
		b = b.Memory(storage.StorageScopeLocal, storage.NewInMemoryStorageProvider())
	}
	return sporecore.NewToolContext(sporecore.SessionID(session), nil, b.Build().Memory())
}

func parseEntries(t *testing.T, out sporecore.ToolOutput) []storage.MemoryEntry {
	t.Helper()
	if out.Kind != sporecore.ToolOutputSuccess {
		t.Fatalf("expected Success, got %+v", out)
	}
	var entries []storage.MemoryEntry
	if err := json.Unmarshal([]byte(out.Content), &entries); err != nil {
		t.Fatalf("unmarshal entries: %v", err)
	}
	return entries
}

func parseEntry(t *testing.T, out sporecore.ToolOutput) storage.MemoryEntry {
	t.Helper()
	if out.Kind != sporecore.ToolOutputSuccess {
		t.Fatalf("expected Success, got %+v", out)
	}
	var e storage.MemoryEntry
	if err := json.Unmarshal([]byte(out.Content), &e); err != nil {
		t.Fatalf("unmarshal entry: %v", err)
	}
	return e
}

func entryContents(entries []storage.MemoryEntry) []string {
	out := make([]string, len(entries))
	for i, e := range entries {
		out[i] = e.Content
	}
	return out
}

// failingMemoryStore always fails, proving storage errors map to a recoverable
// tool error (R9). Mirrors the Rust FailingMemoryStore double.
type failingMemoryStore struct{}

func (failingMemoryStore) AppendMemory(context.Context, storage.StorageScope, storage.SessionID, storage.MemoryEntry) error {
	return errors.New("boom")
}
func (failingMemoryStore) GetMemories(context.Context, storage.StorageScope, storage.SessionID, int) ([]storage.MemoryEntry, error) {
	return nil, errors.New("boom")
}
func (f failingMemoryStore) GetMemoriesMerged(ctx context.Context, sessionID storage.SessionID, limit int) ([]storage.MemoryEntry, error) {
	return storage.MergeMemories(ctx, f, sessionID, limit)
}

var _ storage.MemoryStore = failingMemoryStore{}

func run(t *testing.T, ctx *sporecore.ToolContext, input map[string]any) sporecore.ToolOutput {
	t.Helper()
	return NewMemoryTool().Execute(context.Background(), memCall(input), sporecore.AllowAllSandbox{}, ctx)
}

// ── R1 + R2: write→read roundtrip; write returns the serialized entry ─────────

func TestMemoryWriteThenReadRoundtrip(t *testing.T) {
	ctx := inMemoryCtx()
	w := run(t, ctx, map[string]any{"operation": "write", "scope": "user", "role": "user", "content": "hello"})
	written := parseEntry(t, w)
	if written.Role != "user" || written.Content != "hello" {
		t.Fatalf("written entry mismatch: %+v", written)
	}
	if string(written.Metadata) != "{}" {
		t.Fatalf("metadata default = %q, want {}", string(written.Metadata))
	}

	r := run(t, ctx, map[string]any{"operation": "read", "scope": "user"})
	entries := parseEntries(t, r)
	if len(entries) != 1 || entries[0].Content != "hello" {
		t.Fatalf("read = %v", entryContents(entries))
	}
}

// ── R4: metadata preserved verbatim ───────────────────────────────────────────

func TestMemoryWritePreservesMetadata(t *testing.T) {
	ctx := inMemoryCtx()
	w := run(t, ctx, map[string]any{
		"operation": "write", "scope": "project", "role": "assistant", "content": "c",
		"metadata": map[string]any{"k": "v", "n": 3},
	})
	got := parseEntry(t, w)
	var md map[string]any
	if err := json.Unmarshal(got.Metadata, &md); err != nil {
		t.Fatalf("metadata: %v", err)
	}
	if md["k"] != "v" || md["n"].(float64) != 3 {
		t.Fatalf("metadata not preserved: %v", md)
	}
}

// ── R5: non-merged scope isolation ────────────────────────────────────────────

func TestMemoryScopedReadIsolation(t *testing.T) {
	ctx := inMemoryCtx()
	run(t, ctx, map[string]any{"operation": "write", "scope": "user", "role": "user", "content": "u1"})
	run(t, ctx, map[string]any{"operation": "write", "scope": "project", "role": "assistant", "content": "p1"})

	u := parseEntries(t, run(t, ctx, map[string]any{"operation": "read", "scope": "user"}))
	if len(u) != 1 || u[0].Content != "u1" {
		t.Fatalf("user read = %v", entryContents(u))
	}
	p := parseEntries(t, run(t, ctx, map[string]any{"operation": "read", "scope": "project"}))
	if len(p) != 1 || p[0].Content != "p1" {
		t.Fatalf("project read = %v", entryContents(p))
	}
}

// ── R6: merged read drives the shared merge fixture ───────────────────────────

func TestMemoryMergedReadFixtureReplay(t *testing.T) {
	b, err := os.ReadFile(filepath.Join(fixturesStorageDir(t), "memory_scoped_merge.json"))
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var f struct {
		Limit    int                   `json:"limit"`
		User     []storage.MemoryEntry `json:"user"`
		Project  []storage.MemoryEntry `json:"project"`
		Local    []storage.MemoryEntry `json:"local"`
		Expected []string              `json:"expected_merged_contents"`
	}
	if err := json.Unmarshal(b, &f); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}

	ctx := scopedCtx("s", true)
	store := ctx.MemoryStore.(storage.MemoryStore)
	seed := func(scope storage.StorageScope, es []storage.MemoryEntry) {
		for _, e := range es {
			if err := store.AppendMemory(context.Background(), scope, ctx.SessionID, e); err != nil {
				t.Fatalf("seed: %v", err)
			}
		}
	}
	seed(storage.StorageScopeUser, f.User)
	seed(storage.StorageScopeProject, f.Project)
	seed(storage.StorageScopeLocal, f.Local)

	out := run(t, ctx, map[string]any{"operation": "read", "scope": "user", "merged": true, "limit": f.Limit})
	contents := entryContents(parseEntries(t, out))
	if len(contents) != len(f.Expected) {
		t.Fatalf("merged = %v want %v", contents, f.Expected)
	}
	for i := range contents {
		if contents[i] != f.Expected[i] {
			t.Fatalf("merged = %v want %v", contents, f.Expected)
		}
	}
	dup := 0
	for _, c := range contents {
		if c == "dup" {
			dup++
		}
		if c == "l-should-not-appear" {
			t.Fatalf("local leaked into merge")
		}
	}
	if dup != 2 {
		t.Fatalf("expected both dup entries, got %d", dup)
	}
}

// ── merged read respects limit ────────────────────────────────────────────────

func TestMemoryMergedReadRespectsLimit(t *testing.T) {
	ctx := scopedCtx("s", false)
	store := ctx.MemoryStore.(storage.MemoryStore)
	for i, ts := range []string{"2026-05-01T00:00:00Z", "2026-05-02T00:00:00Z", "2026-05-03T00:00:00Z"} {
		e := storage.NewMemoryEntry("user", string(rune('0'+i)), storage.Timestamp(ts))
		e.Content = "u" + string(rune('0'+i))
		if err := store.AppendMemory(context.Background(), storage.StorageScopeUser, ctx.SessionID, e); err != nil {
			t.Fatalf("append: %v", err)
		}
	}
	out := run(t, ctx, map[string]any{"operation": "read", "scope": "user", "merged": true, "limit": 2})
	entries := parseEntries(t, out)
	if len(entries) != 2 || entries[0].Content != "u2" || entries[1].Content != "u1" {
		t.Fatalf("limited merged read = %v", entryContents(entries))
	}
}

// ── R7: Local rejected on write — exact message, nothing written ──────────────

func TestMemoryLocalRejectedOnWriteWritesNothing(t *testing.T) {
	ctx := scopedCtx("s", true)
	out := run(t, ctx, map[string]any{"operation": "write", "scope": "local", "role": "user", "content": "x"})
	if out.Kind != sporecore.ToolOutputError || !out.Recoverable {
		t.Fatalf("expected recoverable error, got %+v", out)
	}
	if out.Message != LocalScopeRejectedMessage {
		t.Fatalf("message = %q want %q", out.Message, LocalScopeRejectedMessage)
	}
	store := ctx.MemoryStore.(storage.MemoryStore)
	for _, scope := range []storage.StorageScope{storage.StorageScopeUser, storage.StorageScopeProject, storage.StorageScopeLocal} {
		got, _ := store.GetMemories(context.Background(), scope, ctx.SessionID, 50)
		if len(got) != 0 {
			t.Fatalf("scope %q should be empty, got %v", scope, entryContents(got))
		}
	}
}

// ── R7: Local rejected on read — exact message ────────────────────────────────

func TestMemoryLocalRejectedOnRead(t *testing.T) {
	ctx := inMemoryCtx()
	out := run(t, ctx, map[string]any{"operation": "read", "scope": "local"})
	if out.Kind != sporecore.ToolOutputError || !out.Recoverable {
		t.Fatalf("expected recoverable error, got %+v", out)
	}
	if out.Message != LocalScopeRejectedMessage {
		t.Fatalf("message = %q want %q", out.Message, LocalScopeRejectedMessage)
	}
}

// ── Session isolation: two sessions over the SAME store ───────────────────────

func TestMemoryKeyedBySessionID(t *testing.T) {
	store := storage.NewInMemoryStorageProvider()
	ctxA := ctxWith(store, "session-a")
	ctxB := ctxWith(store, "session-b")
	run(t, ctxA, map[string]any{"operation": "write", "scope": "user", "role": "user", "content": "a1"})
	run(t, ctxB, map[string]any{"operation": "write", "scope": "user", "role": "user", "content": "b1"})

	a, _ := store.GetMemories(context.Background(), storage.StorageScopeUser, "session-a", 50)
	b, _ := store.GetMemories(context.Background(), storage.StorageScopeUser, "session-b", 50)
	if len(a) != 1 || a[0].Content != "a1" {
		t.Fatalf("session-a = %v", entryContents(a))
	}
	if len(b) != 1 || b[0].Content != "b1" {
		t.Fatalf("session-b = %v", entryContents(b))
	}
}

// ── R8: bad params → recoverable error ────────────────────────────────────────

func TestMemoryBadParamsRecoverable(t *testing.T) {
	ctx := inMemoryCtx()
	// Unknown operation (scope present so scope check passes; op check fails).
	out := run(t, ctx, map[string]any{"operation": "nope", "scope": "user"})
	if out.Kind != sporecore.ToolOutputError || !out.Recoverable {
		t.Fatalf("unknown op: %+v", out)
	}
	// Missing required field on write (no content).
	out = run(t, ctx, map[string]any{"operation": "write", "scope": "user", "role": "user"})
	if out.Kind != sporecore.ToolOutputError || !out.Recoverable {
		t.Fatalf("missing content: %+v", out)
	}
	// Missing scope.
	out = run(t, ctx, map[string]any{"operation": "read"})
	if out.Kind != sporecore.ToolOutputError || !out.Recoverable {
		t.Fatalf("missing scope: %+v", out)
	}
	// Malformed JSON input.
	bad := sporecore.ToolCall{ID: "c", Name: MemoryToolName, Input: json.RawMessage(`{`)}
	out = NewMemoryTool().Execute(context.Background(), bad, sporecore.AllowAllSandbox{}, ctx)
	if out.Kind != sporecore.ToolOutputError || !out.Recoverable {
		t.Fatalf("malformed: %+v", out)
	}
}

// ── R9: storage failure → recoverable error (both write and read) ─────────────

func TestMemoryStorageFailureRecoverable(t *testing.T) {
	ctx := ctxWith(failingMemoryStore{}, "test-session")
	w := run(t, ctx, map[string]any{"operation": "write", "scope": "user", "role": "user", "content": "x"})
	if w.Kind != sporecore.ToolOutputError || !w.Recoverable {
		t.Fatalf("write failure: %+v", w)
	}
	r := run(t, ctx, map[string]any{"operation": "read", "scope": "user"})
	if r.Kind != sporecore.ToolOutputError || !r.Recoverable {
		t.Fatalf("read failure: %+v", r)
	}
	m := run(t, ctx, map[string]any{"operation": "read", "scope": "user", "merged": true})
	if m.Kind != sporecore.ToolOutputError || !m.Recoverable {
		t.Fatalf("merged read failure: %+v", m)
	}
}

// ── R10: read does not write ──────────────────────────────────────────────────

func TestMemoryReadDoesNotWrite(t *testing.T) {
	ctx := inMemoryCtx()
	r := run(t, ctx, map[string]any{"operation": "read", "scope": "user"})
	if len(parseEntries(t, r)) != 0 {
		t.Fatalf("fresh read should be empty")
	}
	store := ctx.MemoryStore.(storage.MemoryStore)
	got, _ := store.GetMemories(context.Background(), storage.StorageScopeUser, ctx.SessionID, 50)
	if len(got) != 0 {
		t.Fatalf("read mutated the store: %v", entryContents(got))
	}
}

// ── Schema is NOT read_only (decision E) ──────────────────────────────────────

func TestMemorySchemaNotReadOnly(t *testing.T) {
	s := NewMemoryTool().Schema()
	if s.Annotations.ReadOnly || s.Annotations.Destructive || s.Annotations.OpenWorld {
		t.Fatalf("memory annotations must all be false: %+v", s.Annotations)
	}
	if s.Name != "memory" {
		t.Fatalf("name = %q", s.Name)
	}
}

// ── Default read limit is 50 (decision B) ─────────────────────────────────────

func TestMemoryReadDefaultLimit50(t *testing.T) {
	ctx := inMemoryCtx()
	for i := 0; i < 60; i++ {
		run(t, ctx, map[string]any{"operation": "write", "scope": "user", "role": "user", "content": "m" + itoaMem(i)})
	}
	r := run(t, ctx, map[string]any{"operation": "read", "scope": "user"})
	if n := len(parseEntries(t, r)); n != 50 {
		t.Fatalf("default-limit read returned %d, want 50", n)
	}
}

// ── Fixture replay — fixtures/tools/memory.json ───────────────────────────────

type memOpStep struct {
	Input    map[string]any `json:"input"`
	Expected struct {
		OK        bool      `json:"ok"`
		Contents  *[]string `json:"contents"`
		Unordered bool      `json:"unordered"`
		Error     *string   `json:"error"`
	} `json:"expected"`
}

type memOpScenario struct {
	Name  string      `json:"name"`
	Steps []memOpStep `json:"steps"`
}

func TestMemoryFixtureReplayOperations(t *testing.T) {
	data, err := os.ReadFile(filepath.Join(fixturesDir(t), "memory.json"))
	if err != nil {
		t.Fatalf("read fixture: %v", err)
	}
	var scenarios []memOpScenario
	if err := json.Unmarshal(data, &scenarios); err != nil {
		t.Fatalf("unmarshal: %v", err)
	}
	if len(scenarios) == 0 {
		t.Fatal("expected >= 1 scenario")
	}

	for _, sc := range scenarios {
		// Fresh isolated scope-routing provider per scenario.
		ctx := scopedCtx("fx", false)
		for i, step := range sc.Steps {
			out := run(t, ctx, step.Input)
			if step.Expected.OK {
				if out.Kind != sporecore.ToolOutputSuccess {
					t.Fatalf("%s step %d: expected ok, got %+v", sc.Name, i, out)
				}
				if step.Expected.Contents == nil {
					continue
				}
				got := entryContents(parseEntries(t, out))
				want := append([]string(nil), *step.Expected.Contents...)
				if step.Expected.Unordered {
					// Honor the `unordered` flag: tool-stamped timestamps can collide
					// within the same wall-clock second, so compare as a multiset.
					sort.Strings(got)
					sort.Strings(want)
				}
				if len(got) != len(want) {
					t.Fatalf("%s step %d: got %v want %v", sc.Name, i, got, want)
				}
				for j := range got {
					if got[j] != want[j] {
						t.Fatalf("%s step %d: got %v want %v", sc.Name, i, got, want)
					}
				}
			} else {
				if out.Kind != sporecore.ToolOutputError {
					t.Fatalf("%s step %d: expected error, got %+v", sc.Name, i, out)
				}
				if !out.Recoverable {
					t.Fatalf("%s step %d: errors must be recoverable", sc.Name, i)
				}
				if step.Expected.Error != nil && out.Message != *step.Expected.Error {
					t.Fatalf("%s step %d: message %q want %q", sc.Name, i, out.Message, *step.Expected.Error)
				}
			}
		}
	}
}

// fixturesStorageDir locates /fixtures/storage relative to this test file.
func fixturesStorageDir(t *testing.T) string {
	t.Helper()
	_, here, _, _ := runtime.Caller(0)
	return filepath.Join(filepath.Dir(here), "..", "..", "..", "fixtures", "storage")
}

func itoaMem(i int) string {
	if i == 0 {
		return "0"
	}
	var b []byte
	for i > 0 {
		b = append([]byte{byte('0' + i%10)}, b...)
		i /= 10
	}
	return string(b)
}

File: typescript/packages/tools/tests/memory.test.ts
/**
 * MemoryTool (#82) — rules R1–R10 plus fixture replays.
 *
 * Mirrors the Rust suite in `rust/crates/spore-core/src/tools/memory.rs`:
 * write→read roundtrip, metadata preservation, scope isolation, merged read
 * (driving `fixtures/storage/memory_scoped_merge.json`), Local rejection on both
 * ops (exact message + nothing written), session isolation, bad-params and
 * storage-error recoverability, read-does-not-write, schema-not-read_only, the
 * default limit of 50, and a step-by-step replay of `fixtures/tools/memory.json`
 * (honoring the per-step `unordered` flag).
 */

import { readFileSync } from "node:fs";
import { dirname, resolve } from "node:path";
import { fileURLToPath } from "node:url";

import {
  memory as coreMemory,
  SessionId,
  storage,
  toolRegistry,
  type SandboxProvider,
  type SandboxViolation,
  type ToolCall,
  type ToolOutput,
} from "@spore/core";
import { describe, expect, it } from "vitest";

import { MEMORY_LOCAL_REJECTED_MESSAGE, MemoryTool } from "../src/index.js";

type MemoryEntry = storage.MemoryEntry;
type MemoryStore = storage.MemoryStore;
type StorageScope = storage.StorageScope;

const { Timestamp } = coreMemory;
const { ToolContext } = toolRegistry;
const { InMemoryStorageProvider, CompositeStorageProvider, ProjectId } = storage;
const TEST_PROJECT = ProjectId.fromCanonicalPath("/test-project");

const here = dirname(fileURLToPath(import.meta.url));
const toolsFixtures = resolve(here, "../../../../fixtures/tools");
const storageFixtures = resolve(here, "../../../../fixtures/storage");

/** Permissive sandbox — the tool never touches the filesystem. */
class AllowAllSandbox implements SandboxProvider {
  async validate(_call: ToolCall): Promise<SandboxViolation | null> {
    return null;
  }
}

/** A MemoryStore that always fails — proves storage errors map to a recoverable error. */
class FailingMemoryStore implements MemoryStore {
  async appendMemory(
    _scope: StorageScope,
    _sessionId: SessionId,
    _entry: MemoryEntry,
  ): Promise<void> {
    throw new Error("boom");
  }
  async getMemories(
    _scope: StorageScope,
    _sessionId: SessionId,
    _limit: number,
  ): Promise<MemoryEntry[]> {
    throw new Error("boom");
  }
  async getMemoriesMerged(
    _sessionId: SessionId,
    _limit: number,
  ): Promise<MemoryEntry[]> {
    throw new Error("boom");
  }
}

function ctxWith(
  memoryStore: MemoryStore,
  session = "test-session",
): toolRegistry.ToolContext {
  // MemoryTool exercises the memory seam only; the run store is a no-op
  // in-memory backend here.
  return new ToolContext(
    SessionId.of(session),
    TEST_PROJECT,
    new InMemoryStorageProvider(),
    memoryStore,
  );
}

function inMemoryCtx(): toolRegistry.ToolContext {
  return ctxWith(new InMemoryStorageProvider());
}

function call(input: unknown): ToolCall {
  return { id: "c1", name: MemoryTool.NAME, input };
}

function parseEntries(out: ToolOutput): MemoryEntry[] {
  if (out.kind !== "success")
    throw new Error(`expected success, got ${JSON.stringify(out)}`);
  return JSON.parse(out.content) as MemoryEntry[];
}

function parseEntry(out: ToolOutput): MemoryEntry {
  if (out.kind !== "success")
    throw new Error(`expected success, got ${JSON.stringify(out)}`);
  return JSON.parse(out.content) as MemoryEntry;
}

const tool = new MemoryTool();
const sb = new AllowAllSandbox();

describe("MemoryTool", () => {
  // R1 + R2: write→read roundtrip; write returns the serialized entry.
  it("R1/R2: write then read roundtrip; write returns the entry", async () => {
    const ctx = inMemoryCtx();
    const w = await tool.execute(
      call({
        operation: "write",
        scope: "user",
        role: "user",
        content: "hello",
      }),
      sb,
      ctx,
    );
    const written = parseEntry(w);
    expect(written.role).toBe("user");
    expect(written.content).toBe("hello");
    expect(written.metadata).toEqual({}); // R4 default {}

    const r = await tool.execute(
      call({ operation: "read", scope: "user" }),
      sb,
      ctx,
    );
    const entries = parseEntries(r);
    expect(entries.length).toBe(1);
    expect(entries[0]!.content).toBe("hello");
  });

  // R4: metadata stored verbatim on the entry.
  it("R4: write preserves metadata", async () => {
    const ctx = inMemoryCtx();
    const w = await tool.execute(
      call({
        operation: "write",
        scope: "project",
        role: "assistant",
        content: "c",
        metadata: { k: "v", n: 3 },
      }),
      sb,
      ctx,
    );
    expect(parseEntry(w).metadata).toEqual({ k: "v", n: 3 });
  });

  // R5: non-merged scope isolation.
  it("R5: a scoped read does not see the other scope", async () => {
    const ctx = inMemoryCtx();
    await tool.execute(
      call({ operation: "write", scope: "user", role: "user", content: "u1" }),
      sb,
      ctx,
    );
    await tool.execute(
      call({
        operation: "write",
        scope: "project",
        role: "assistant",
        content: "p1",
      }),
      sb,
      ctx,
    );

    const user = parseEntries(
      await tool.execute(call({ operation: "read", scope: "user" }), sb, ctx),
    );
    expect(user.map((e) => e.content)).toEqual(["u1"]);

    const proj = parseEntries(
      await tool.execute(
        call({ operation: "read", scope: "project" }),
        sb,
        ctx,
      ),
    );
    expect(proj.map((e) => e.content)).toEqual(["p1"]);
  });

  // R6: merged read drives the shared merge fixture — both `dup` survive,
  // `local` absent, newest-first.
  it("R6: merged read replays memory_scoped_merge.json", async () => {
    const f = JSON.parse(
      readFileSync(
        resolve(storageFixtures, "memory_scoped_merge.json"),
        "utf8",
      ),
    ) as {
      limit: number;
      user: MemoryEntry[];
      project: MemoryEntry[];
      local: MemoryEntry[];
      expected_merged_contents: string[];
    };

    const provider = new CompositeStorageProvider()
      .memory("user", new InMemoryStorageProvider())
      .memory("project", new InMemoryStorageProvider())
      .memory("local", new InMemoryStorageProvider())
      .build();
    const memoryStore = provider.memory();
    const sid = SessionId.of("s");
    for (const [key, scope] of [
      ["user", "user"],
      ["project", "project"],
      ["local", "local"],
    ] as const) {
      for (const raw of f[key]) {
        await memoryStore.appendMemory(scope, sid, {
          role: raw.role,
          content: raw.content,
          timestamp: Timestamp.of(raw.timestamp as unknown as string),
          metadata: raw.metadata,
        });
      }
    }

    const ctx = new ToolContext(
      sid,
      TEST_PROJECT,
      new InMemoryStorageProvider(),
      memoryStore,
    );
    const out = await tool.execute(
      call({ operation: "read", scope: "user", merged: true, limit: f.limit }),
      sb,
      ctx,
    );
    const contents = parseEntries(out).map((e) => e.content);
    expect(contents).toEqual(f.expected_merged_contents);
    expect(contents.filter((c) => c === "dup").length).toBe(2);
    expect(contents.some((c) => c.includes("should-not-appear"))).toBe(false);
  });

  // merged read respects limit: 3 user entries, merged read limit=2.
  it("merged read respects limit (newest-first)", async () => {
    const provider = new CompositeStorageProvider()
      .memory("user", new InMemoryStorageProvider())
      .memory("project", new InMemoryStorageProvider())
      .build();
    const memoryStore = provider.memory();
    const sid = SessionId.of("s");
    const stamps = [
      "2026-05-01T00:00:00Z",
      "2026-05-02T00:00:00Z",
      "2026-05-03T00:00:00Z",
    ];
    for (let i = 0; i < stamps.length; i++) {
      await memoryStore.appendMemory("user", sid, {
        role: "user",
        content: `u${i}`,
        timestamp: Timestamp.of(stamps[i]!),
        metadata: {},
      });
    }
    const ctx = new ToolContext(
      sid,
      TEST_PROJECT,
      new InMemoryStorageProvider(),
      memoryStore,
    );
    const out = await tool.execute(
      call({ operation: "read", scope: "user", merged: true, limit: 2 }),
      sb,
      ctx,
    );
    const entries = parseEntries(out);
    expect(entries.map((e) => e.content)).toEqual(["u2", "u1"]);
  });

  // R7: Local rejected on write — exact message, nothing written.
  it("R7: local rejected on write writes nothing", async () => {
    const ctx = inMemoryCtx();
    const out = await tool.execute(
      call({ operation: "write", scope: "local", role: "user", content: "x" }),
      sb,
      ctx,
    );
    expect(out.kind).toBe("error");
    if (out.kind === "error") {
      expect(out.recoverable).toBe(true);
      expect(out.message).toBe(MEMORY_LOCAL_REJECTED_MESSAGE);
    }
    // Nothing was written to ANY scope.
    for (const scope of ["user", "project", "local"] as const) {
      const got = await ctx.memoryStore.getMemories(scope, ctx.sessionId, 50);
      expect(got).toEqual([]);
    }
  });

  // R7: Local rejected on read — exact message.
  it("R7: local rejected on read", async () => {
    const ctx = inMemoryCtx();
    const out = await tool.execute(
      call({ operation: "read", scope: "local" }),
      sb,
      ctx,
    );
    expect(out.kind).toBe("error");
    if (out.kind === "error") {
      expect(out.recoverable).toBe(true);
      expect(out.message).toBe(MEMORY_LOCAL_REJECTED_MESSAGE);
    }
  });

  // Session isolation: two sessions over the SAME store keep separate memory.
  it("memory is keyed by session id", async () => {
    const store = new InMemoryStorageProvider();
    const ctxA = ctxWith(store, "session-a");
    const ctxB = ctxWith(store, "session-b");

    await tool.execute(
      call({ operation: "write", scope: "user", role: "user", content: "a1" }),
      sb,
      ctxA,
    );
    await tool.execute(
      call({ operation: "write", scope: "user", role: "user", content: "b1" }),
      sb,
      ctxB,
    );

    const a = await store.getMemories("user", SessionId.of("session-a"), 50);
    const b = await store.getMemories("user", SessionId.of("session-b"), 50);
    expect(a.map((e) => e.content)).toEqual(["a1"]);
    expect(b.map((e) => e.content)).toEqual(["b1"]);
  });

  // R8: bad params → recoverable error.
  it("R8: bad params are a recoverable error", async () => {
    const ctx = inMemoryCtx();
    // Unknown operation.
    const r1 = await tool.execute(call({ operation: "nope" }), sb, ctx);
    expect(r1.kind).toBe("error");
    if (r1.kind === "error") expect(r1.recoverable).toBe(true);
    // Missing required field on write.
    const r2 = await tool.execute(
      call({ operation: "write", scope: "user", role: "user" }),
      sb,
      ctx,
    );
    expect(r2.kind).toBe("error");
    if (r2.kind === "error") expect(r2.recoverable).toBe(true);
  });

  // R9: storage failure → recoverable error (both write and read paths).
  it("R9: storage failure is a recoverable error", async () => {
    const ctx = ctxWith(new FailingMemoryStore());
    const w = await tool.execute(
      call({ operation: "write", scope: "user", role: "user", content: "x" }),
      sb,
      ctx,
    );
    expect(w.kind).toBe("error");
    if (w.kind === "error") expect(w.recoverable).toBe(true);

    const r = await tool.execute(
      call({ operation: "read", scope: "user" }),
      sb,
      ctx,
    );
    expect(r.kind).toBe("error");
    if (r.kind === "error") expect(r.recoverable).toBe(true);
  });

  // R10: read does not write — a read against a fresh store leaves it empty.
  it("R10: read does not write", async () => {
    const ctx = inMemoryCtx();
    const r = await tool.execute(
      call({ operation: "read", scope: "user" }),
      sb,
      ctx,
    );
    expect(parseEntries(r)).toEqual([]);
    const got = await ctx.memoryStore.getMemories("user", ctx.sessionId, 50);
    expect(got).toEqual([]);
  });

  // Schema is NOT read_only (decision E).
  it("schema is not read_only", () => {
    const s = MemoryTool.schema();
    expect(s.annotations.read_only).toBe(false);
    expect(s.annotations.destructive).toBe(false);
    expect(s.annotations.open_world).toBe(false);
    expect(s.name).toBe("memory");
    // Advertised scope enum omits `local`.
    const props = (
      s.parameters as { properties: { scope: { enum: string[] } } }
    ).properties;
    expect(props.scope.enum).toEqual(["project", "user"]);
  });

  // read default limit is 50 (decision B): more than 50 entries → 50 returned.
  it("R3: read default limit is 50", async () => {
    const ctx = inMemoryCtx();
    for (let i = 0; i < 60; i++) {
      await tool.execute(
        call({
          operation: "write",
          scope: "user",
          role: "user",
          content: `m${i}`,
        }),
        sb,
        ctx,
      );
    }
    const r = await tool.execute(
      call({ operation: "read", scope: "user" }),
      sb,
      ctx,
    );
    expect(parseEntries(r).length).toBe(50);
  });
});

// ============================================================================
// Fixture replay — fixtures/tools/memory.json
// ============================================================================

interface OpExpected {
  ok: boolean;
  contents?: string[];
  unordered?: boolean;
  error?: string;
}
interface OpStep {
  input: unknown;
  expected: OpExpected;
}
interface OpScenario {
  name: string;
  steps: OpStep[];
}

describe("MemoryTool fixture replay (memory.json)", () => {
  const scenarios = JSON.parse(
    readFileSync(resolve(toolsFixtures, "memory.json"), "utf8"),
  ) as OpScenario[];

  it("loads at least one scenario", () => {
    expect(scenarios.length).toBeGreaterThan(0);
  });

  for (const sc of scenarios) {
    it(`replays ${sc.name}`, async () => {
      // Fresh isolated scope-routing provider per scenario.
      const provider = new CompositeStorageProvider()
        .memory("user", new InMemoryStorageProvider())
        .memory("project", new InMemoryStorageProvider())
        .build();
      const ctx = new ToolContext(
        SessionId.of("fx"),
        TEST_PROJECT,
        new InMemoryStorageProvider(),
        provider.memory(),
      );

      for (let i = 0; i < sc.steps.length; i++) {
        const step = sc.steps[i]!;
        const out = await tool.execute(call(step.input), sb, ctx);
        if (step.expected.ok) {
          expect(out.kind, `${sc.name} step ${i}`).toBe("success");
          // A read asserts its content list; a write only asserts ok.
          if (step.expected.contents) {
            const got = parseEntries(out).map((e) => e.content);
            if (step.expected.unordered) {
              expect([...got].sort(), `${sc.name} step ${i}`).toEqual(
                [...step.expected.contents].sort(),
              );
            } else {
              expect(got, `${sc.name} step ${i}`).toEqual(
                step.expected.contents,
              );
            }
          }
        } else {
          expect(out.kind, `${sc.name} step ${i}`).toBe("error");
          if (out.kind === "error") {
            expect(
              out.recoverable,
              `${sc.name} step ${i}: errors must be recoverable`,
            ).toBe(true);
            expect(out.message, `${sc.name} step ${i}`).toBe(
              step.expected.error,
            );
          }
        }
      }
    });
  }
});
